### ISSUE TITLE
Client IMap errors on Imap with backups while shutdown of remote member node 3.8+

### ISSUE THREAD

**Issue by agseijas** [hazelcast/hazelcast#13396]
Hi,

I've just created [this repository](https://github.com/agseijas/hazelcast-client-errors-shutdown-test) containing the [src/main/test/hz/test/ClientWithShuttingDownRemoteServersIT.java test](https://github.com/agseijas/hazelcast-client-errors-shutdown-test/blob/master/src/main/test/hz/test/ClientWithShuttingDownRemoteServersIT.java). Test is using Spring just to ease the setup.

The test scenario is composed of a client instance connected to two member nodes (master + 1) in clustered mode with 1 async backup on the cache being used. The client performs continuously cache operations, so that it fills the cache up to 1.5GB and then when it reaches that point, the slave node is shutdown, producing many errors on client side on non-ready-only ops. (puts) because the dying slave is passive.

If there were no backups I was expecting the operations to fail and that's ok because of the passive mode on the shutting down node. But with the backups I wasn't expecting any, moreover, I couldn't find any way to resolve this 100% (ie: no errors while keeping the provided fast-response/fail configuration). Note that this is just a test based on a real scenario in production with real member nodes and real client nodes in different metal, with much much higher concurrency (20k put op/s) the number of errors we are having there are massive while putting a server down.

Feel free to tamper with the test code as its setup is really complex and might fail on your own setup. Do not hesitate to contact me if something is wrong or needs fixing/explaining on the test.

NOTE: I understand that retries with small timeouts are the ones producing the errors, also I noticed strangely that on version 3.10.2 of Hz there are a lot more errors happening.

**Comment by ihsandemir** [comment]
@agseijas Looking at your instance configuration, I see that you configured the number of async backups as 1. Why do you think that the second member is passive? Map data is partitioned into the two members, and the backup of a partition on one server will be the other server and vice versa. See https://docs.hazelcast.org/docs/3.11.2/manual/html-single/index.html#data-partitioning for an example how the backups are replicated and distributed over multiple members. All members are actually active in the sense that they own some partitions. Does this make sense?

**Comment by agseijas** [comment]
Sorry, didn't express myself probably, of course the cluster member is active (I'm using the passive/active wording as defined per cluster member state in here https://docs.hazelcast.org/docs/3.11.2/manual/html-single/index.html#cluster-member-states); the passive mode activates (and I'm probably assuming this as well) is when an instance is shutdown during the test execution time.

Nevertheless, you can also see this PASSIVE state per instance on the exception, which is also the one you're asking:

`com.hazelcast.core.OperationTimeoutException: ClientInvocation{clientMessage = ClientMessage{length=22, correlationId=34181, operation=null, messageType=112, partitionId=134, isComplete=false, isRetryable=false, isEvent=false, writeOffset=0}, objectName = myRemoteCache, target = partition 134, sendConnection = ClientConnection{alive=true, connectionId=2, channel=NioChannel{/10.28.61.219:49755->/10.28.61.219:11112}, remoteEndpoint=[10.28.61.219]:11112, lastReadTime=2019-03-20 17:38:52.937, lastWriteTime=2019-03-20 17:38:52.936, closedTime=never, lastHeartbeatRequested=2019-03-20 17:38:52.936, lastHeartbeatReceived=2019-03-20 17:38:52.936, connected server version=3.10.2}} timed out because exception occurred after client invocation timeout 1000 ms. Current time: 2019-03-20 17:38:52.937. Start time: 2019-03-20 17:38:51.350. Total elapsed time: 1587 ms. 
	at com.hazelcast.client.spi.impl.ClientInvocation.newOperationTimeoutException(ClientInvocation.java:344)
	at com.hazelcast.client.spi.impl.ClientInvocation.notifyException(ClientInvocation.java:241)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.handleClientMessage(ClientResponseHandlerSupplier.java:134)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.process(ClientResponseHandlerSupplier.java:115)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.access$500(ClientResponseHandlerSupplier.java:50)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier$ResponseThread.doRun(ClientResponseHandlerSupplier.java:168)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier$ResponseThread.run(ClientResponseHandlerSupplier.java:152)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:162)
	at com.hazelcast.client.spi.ClientProxy.invokeOnPartition(ClientProxy.java:225)
	at com.hazelcast.client.spi.ClientProxy.invoke(ClientProxy.java:219)
	at com.hazelcast.client.proxy.ClientMapProxy.setInternal(ClientMapProxy.java:612)
	at com.hazelcast.client.proxy.ClientMapProxy.set(ClientMapProxy.java:603)
	at com.hazelcast.client.proxy.ClientMapProxy.set(ClientMapProxy.java:1522)
	at com.hazelcast.spring.cache.HazelcastCache.put(HazelcastCache.java:114)
	at hz.test.ClientWithShuttingDownRemoteServersIT.performPutAndReport(ClientWithShuttingDownRemoteServersIT.java:163)
	at hz.test.ClientWithShuttingDownRemoteServersIT.lambda$putInfinitelyUntilItBreaks$1(ClientWithShuttingDownRemoteServersIT.java:150)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.core.HazelcastInstanceNotActiveException: State: PASSIVE Operation: class com.hazelcast.map.impl.operation.SetOperation
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.engineActive(Invocation.java:502)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:535)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:525)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:215)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:60)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:67)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.doRun(AbstractMessageTask.java:111)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:101)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:155)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:127)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:250)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.execute(AbstractPartitionMessageTask.java:78)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.unblock(AbstractInvocationFuture.java:246)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.andThen(AbstractInvocationFuture.java:222)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:69)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.doRun(AbstractMessageTask.java:111)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:101)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:155)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
com.hazelcast.core.OperationTimeoutException: ClientInvocation{clientMessage = ClientMessage{length=22, correlationId=34189, operation=null, messageType=112, partitionId=79, isComplete=false, isRetryable=false, isEvent=false, writeOffset=0}, objectName = myRemoteCache, target = partition 79, sendConnection = ClientConnection{alive=true, connectionId=2, channel=NioChannel{/10.28.61.219:49755->/10.28.61.219:11112}, remoteEndpoint=[10.28.61.219]:11112, lastReadTime=2019-03-20 17:38:52.937, lastWriteTime=2019-03-20 17:38:52.936, closedTime=never, lastHeartbeatRequested=2019-03-20 17:38:52.936, lastHeartbeatReceived=2019-03-20 17:38:52.936, connected server version=3.10.2}} timed out because exception occurred after client invocation timeout 1000 ms. Current time: 2019-03-20 17:38:52.937. Start time: 2019-03-20 17:38:51.350. Total elapsed time: 1587 ms. 
	at com.hazelcast.client.spi.impl.ClientInvocation.newOperationTimeoutException(ClientInvocation.java:344)
	at com.hazelcast.client.spi.impl.ClientInvocation.notifyException(ClientInvocation.java:241)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.handleClientMessage(ClientResponseHandlerSupplier.java:134)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.process(ClientResponseHandlerSupplier.java:115)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.access$500(ClientResponseHandlerSupplier.java:50)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier$ResponseThread.doRun(ClientResponseHandlerSupplier.java:168)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier$ResponseThread.run(ClientResponseHandlerSupplier.java:152)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:162)
	at com.hazelcast.client.spi.ClientProxy.invokeOnPartition(ClientProxy.java:225)
	at com.hazelcast.client.spi.ClientProxy.invoke(ClientProxy.java:219)
	at com.hazelcast.client.proxy.ClientMapProxy.setInternal(ClientMapProxy.java:612)
	at com.hazelcast.client.proxy.ClientMapProxy.set(ClientMapProxy.java:603)
	at com.hazelcast.client.proxy.ClientMapProxy.set(ClientMapProxy.java:1522)
	at com.hazelcast.spring.cache.HazelcastCache.put(HazelcastCache.java:114)
	at hz.test.ClientWithShuttingDownRemoteServersIT.performPutAndReport(ClientWithShuttingDownRemoteServersIT.java:163)
	at hz.test.ClientWithShuttingDownRemoteServersIT.lambda$putInfinitelyUntilItBreaks$1(ClientWithShuttingDownRemoteServersIT.java:150)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.core.HazelcastInstanceNotActiveException: State: PASSIVE Operation: class com.hazelcast.map.impl.operation.SetOperation
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.engineActive(Invocation.java:502)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:535)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:525)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:215)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:60)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:67)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.doRun(AbstractMessageTask.java:111)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:101)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:155)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:127)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:250)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.execute(AbstractPartitionMessageTask.java:78)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.unblock(AbstractInvocationFuture.java:246)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.andThen(AbstractInvocationFuture.java:222)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:69)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.doRun(AbstractMessageTask.java:111)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:101)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:155)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
com.hazelcast.core.OperationTimeoutException: ClientInvocation{clientMessage = ClientMessage{length=22, correlationId=34188, operation=null, messageType=112, partitionId=268, isComplete=false, isRetryable=false, isEvent=false, writeOffset=0}, objectName = myRemoteCache, target = partition 268, sendConnection = ClientConnection{alive=true, connectionId=2, channel=NioChannel{/10.28.61.219:49755->/10.28.61.219:11112}, remoteEndpoint=[10.28.61.219]:11112, lastReadTime=2019-03-20 17:38:52.937, lastWriteTime=2019-03-20 17:38:52.936, closedTime=never, lastHeartbeatRequested=2019-03-20 17:38:52.936, lastHeartbeatReceived=2019-03-20 17:38:52.936, connected server version=3.10.2}} timed out because exception occurred after client invocation timeout 1000 ms. Current time: 2019-03-20 17:38:52.939. Start time: 2019-03-20 17:38:51.350. Total elapsed time: 1589 ms. 
	at com.hazelcast.client.spi.impl.ClientInvocation.newOperationTimeoutException(ClientInvocation.java:344)
	at com.hazelcast.client.spi.impl.ClientInvocation.notifyException(ClientInvocation.java:241)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.handleClientMessage(ClientResponseHandlerSupplier.java:134)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.process(ClientResponseHandlerSupplier.java:115)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier.access$500(ClientResponseHandlerSupplier.java:50)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier$ResponseThread.doRun(ClientResponseHandlerSupplier.java:168)
	at com.hazelcast.client.spi.impl.ClientResponseHandlerSupplier$ResponseThread.run(ClientResponseHandlerSupplier.java:152)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:162)
	at com.hazelcast.client.spi.ClientProxy.invokeOnPartition(ClientProxy.java:225)
	at com.hazelcast.client.spi.ClientProxy.invoke(ClientProxy.java:219)
	at com.hazelcast.client.proxy.ClientMapProxy.setInternal(ClientMapProxy.java:612)
	at com.hazelcast.client.proxy.ClientMapProxy.set(ClientMapProxy.java:603)
	at com.hazelcast.client.proxy.ClientMapProxy.set(ClientMapProxy.java:1522)
	at com.hazelcast.spring.cache.HazelcastCache.put(HazelcastCache.java:114)
	at hz.test.ClientWithShuttingDownRemoteServersIT.performPutAndReport(ClientWithShuttingDownRemoteServersIT.java:163)
	at hz.test.ClientWithShuttingDownRemoteServersIT.lambda$putInfinitelyUntilItBreaks$1(ClientWithShuttingDownRemoteServersIT.java:150)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.core.HazelcastInstanceNotActiveException: State: PASSIVE Operation: class com.hazelcast.map.impl.operation.SetOperation
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.engineActive(Invocation.java:502)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:535)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:525)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:215)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:60)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:67)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.doRun(AbstractMessageTask.java:111)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:101)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:155)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:127)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:250)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.execute(AbstractPartitionMessageTask.java:78)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.unblock(AbstractInvocationFuture.java:246)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.andThen(AbstractInvocationFuture.java:222)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:69)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.doRun(AbstractMessageTask.java:111)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:101)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:155)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)`

**Comment by ihsandemir** [comment]
@agseijas According to your logs, you received operation timeout exception after 1.5 seconds. HazelcastInstanceNotActiveException exceptions are retried even for Imap.put operations, but until the timeout. In you code, you set the timeout to 1 second:
`clientConfig.setProperty("hazelcast.client.invocation.timeout.seconds", "1");`

And hence, you did not give enough time for recovery. I suggest that you set this to a higher value. The default is 120 seconds if you did not set anything. Please try this and let us know how it works for you.

Furthermore, you are setting the heartbeat timeout very short to 
`clientConfig.setProperty("hazelcast.client.heartbeat.timeout", "2000");`. This is also very short, why did you set this? You may not need to set it at all in most cases and just use the default.

**Comment by agseijas** [comment]
I think that we are missing the point, the timings are just worsening the issue for sure, that's what they're meant for, but IMHO, for a proper, ordered shutdown we should receive no errors client side (or at least a fast and clean recovery), whatever the timings you setup (if they are, like in this case, correct). But that is not the case hence the issue.

In any case as I said originally such values were setup to be able to fail fast and promote a fast recovery.

**Comment by ihsandemir** [comment]
@agseijas Can you try the case with high timeouts (or you can even use the defaults) and let us know what is the exception you are getting? With enough time we do not expect an exception, and if you are seeing an exception even with high timeouts, then we need to examine it. Can you please try it and let us know?

Here is what I expect when a member is shutdown:

1. Client receives `HazelcastInstanceNotActiveException`. Client retries the invocation internally when the `HazelcastInstanceNotActiveException` is received.
2. Client waits for partition table update (this may be a few seconds) after the server is shutdown, the partitions are redistributed over to the alive member and client gets notified.
3. After partition update, the client can now know that the owner of the partition is the alive member and tries the invocation to this member and succeeds.

**Comment by ihsandemir** [comment]
@agseijas Did you get a chance to try above recommendation?

**Comment by agseijas** [comment]
Hi! the point is trying with such values, so that the client timeout's before the migrations finish, that way we can produce the errors. With the default value set so high (ie: higher than the migration time), we risk blocking for an enormous amount of time when using large clusters (specially when rolling upgrades). So yes, testing with your proposal makes no sense as, obviously, it works as expected! I think that you might be expecting a bug, but this, IMHO, is a misbehaviour of the client-server relationship.

Please consider this:

- A client connects to various servers. These clients perform big amounts IMap.getAll operations with high traffic and concurrency that are performed cluster wide.
- The servers hold a large amount of data that is backed up in other servers
- Migrations might take a large amount of time (> client timeouts)
- One server is gracefully shutdown.

And my expectation for clients in this scenario is that clients should recognize this shutdown action, and should behave in the most reliable way: perhaps by shutdown server allowing reads or reading for backup and not blocking for the whole migration time the cluster-wide operations.

**Comment by sancar** [comment]
Hi @agseijas, 

1. The failure we see is related to the shutdown of a member, not migration. Set operation will go to other member and return successfully if the timeout was longer. 

2. During migration read operations are not blocked. Moreover, migration is done in partition level, so if your write operation is not related to the migrating partition, it will not wait. 

3. `client.invocation.timeout.seconds` is not a hard value. Even if you set it 1 second, your put operation will wait for migration to wait. This value is checked only if a failure happens. 
You can refer to here for details https://docs.hazelcast.org/docs/latest/manual/html-single/index.html#java-client-operation-modes

client.invocation.timeout.seconds one second, you should be prepared to get an exception when an exception occurs. If the exception is not safe to retry, we will delegate it to the user anyway. ( default behavior, more specifically redo-operation is not set)

`client.invocation.timeout.seconds`, This setting is related to what do you want an exception occurs that can be handled gracefully. Do you want it to be handled and wait, or do you want the exception to delegated to you?

**Comment by agseijas** [comment]
Hi @sancar 
my expectation is for the client to never block any operation during the whole process. If the shutdown process is handled cleanly:  why should it incur into waits? 
As this is related to shutdown why didn't the client just performed the set operation on the `new` partition already (client reacting to a PASSIVE/SHUTDOWN status on the other end). As the test is creating a 0 latency scenario, I see no further reason for having to use any kind of timeout.
And this leads to the last question: Why is the set operation working only if the timeouts is longer? is there any strong reason for it that's not meant for network outages?.
 
Our app is bound into slow latencies and very high throughputs, so hence the small timeouts (I refer to my first comment's note at the bottom though); currently we catch the given exceptions and ignore them (hence going straight to the underlying repository); so it's not a big deal anymore. In this regard, I wouldn't change the current behaviour of the client.

**Comment by sancar** [comment]
The graceful shutdown process takes some time as you can imagine. You either be willing to wait a couple of seconds, or you don't. The operations took less than a second when everything is ok, but when there are cluster changes, it is expected to see some operations took longer. 

It was going to send the operation to the new partition, the detection of member shutdown and client getting the new partition table took longer than a second. 
There are multiple agents that need to decide the member is shut down. Client cannot decide the cluster view itself, it needs to get that info from another member. A member deciding this member is shut down, updating the partition table and delegating that info to the client. There are no migrations to wait in this scenario, but it is not less than a second as you can imagine. 

The decision is up to you with config. You either wait for the system to handle this gracefully, or you don't and handle it differently. 

We are always investigating the reasons that can stall the client unnecessarily. Optimizing downtime, when a member failure, is one of our constant goals. We did quite a work on that in version 3.11. 
If you report that client is taking longer than expected, we could investigate. In your test, we don't know if it is taking 1.01 second or more. 

Closing the issue for now. Github is used for tracking known bugs. If you can post your similar questions on googlegroup https://groups.google.com/forum/#!forum/hazelcast , more people will see, answer and get involved. If we find a bug to fix, we can always open an issue here later.

**Comment by agseijas** [comment]
I'm sorry in the contributing doc it says `For an enhancement or larger feature, you can create a GitHub issue first to discuss.`, I'm sorry if I misundertood what're github issues for.

Nevermind, I thought that as Hz act as a distributed state, then it can distribute it's own status to prevent service hiccups. IMHO this is a unnecessary client stall, but it's just my opinion; but anyways, we already covered our backs by ignoring the exceptions.

**Comment by sancar** [comment]
@agseijas Yes, it will distribute the state but this distribution itself is the place where time is necessary. Please, if you think that the client is stalling longer than the necessary, continue sharing your findings.
